### PR TITLE
Refactor/categories

### PR DESCRIPTION
**Pull request by JanMaxime**
- Categories now come from a JSON file
- Categories can also be mocked for testing purposes
- The fact categories are mocked for testing purposes is also the cause of bad coverage.

### CODE REVIEW

File: app/src/main/java/ch/epfl/polybazaar/MainActivity.java [hunk 1]
```diff
@@ -19,11 +23,14 @@
...
     private AppUser user;
 
+    @RequiresApi(api = Build.VERSION_CODES.N)
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
         authenticator = AuthenticatorFactory.getDependency();
+        Category c = RootCategory.getInstance(this);
+        System.out.println(c.toString());
```
[the remark this answers is not in the document]
Comment (JanMaxime) [reply app/src/main/java/ch/epfl/polybazaar/MainActivity.java:33]: Good catch, it was for testing purpose and thought I had deleted it, but it must have come back when I reverted some changes. I will delete it

File: app/src/main/java/ch/epfl/polybazaar/category/RootCategory.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        return root;
+    }
+
+    private RootCategory(Context context) throws IOException, JSONException {
+        nodes = new ArrayList<>();
+            BufferedReader reader = new BufferedReader(new InputStreamReader(context.getAssets().open("categories.json")));
+            String line = reader.readLine();
+            StringBuilder jsonBuilder = new StringBuilder();
+
+            while(line != null){
```
[the remark this answers is not in the document]
Comment (JanMaxime) [reply app/src/main/java/ch/epfl/polybazaar/category/RootCategory.java:49]: Will apply !

File: app/src/main/java/ch/epfl/polybazaar/category/RootCategory.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        return size;
+    }
+
+    @Override
+    public int maxDepth(){
+        int maxDepth = 0;
+        for(Category category : nodes){
+            maxDepth = max(maxDepth, category.maxDepth());
+        }
+        return maxDepth;
```
Comment (Walterieux) [inline app/src/main/java/ch/epfl/polybazaar/category/RootCategory.java:126]: `return Collections.max(nodes, Comparator.comparing(c -> c.maxDepth()));`

Comment (JanMaxime) [reply app/src/main/java/ch/epfl/polybazaar/category/RootCategory.java:126]: Sadly the comparing function requires API version 24 at least which causes an issue with Travis

**Pull request merged**